<a href="https://colab.research.google.com/github/michalis0/MGT-502-Data-Science-and-Machine-Learning/blob/main/assignments/Part 1/Assignment_part_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DSML investigation:
### You are part of the Suisse Impossible Mission Force, or SIMF for short. You need to uncover a rogue agent that is trying to steal sensitive information.

### Your mission, should you choose to accept it, is to find that agent before any classified information gets stolen. Good luck!


## Assignment part one

Airport security confiscated a laptop of a spy, which was later given to you. The laptop contains some documents that might be useful in the investigation process. You retrieve those documents and want to analyse them. Our intelligence shows that the person that we are looking for has visited the USA between Sept 2019 to Oct 2020. He is currently working undercover.


### Getting to know our data

We have retrieved from the laptop the following file:
- A list of suspects
- The flight records of these potential suspects



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [32]:
user_records = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%201/data/userRecords.csv")
travel_records = pd.read_csv("https://raw.githubusercontent.com/michalis0/MGT-502-Data-Science-and-Machine-Learning/refs/heads/main/assignments/Part%201/data/flightRecords.csv")

#### Shape of the data

Let's first check how many rows and columns (features) are in the user records

In [4]:
#Your code here
user_records.shape

(957, 8)

**Q1. How many rows are there in the user records dataset?**

There are 957 rows in the user records dataset.


#### Check out the first few rows
Print the first few rows of the user records dataset and check them. Note that to protect innocent people, the name of the suspects have been censored, those will be revealed once number of potential suspects decreases.


In [5]:
#Your code here
user_records.head()

UserID first_name  last_name  birthday country marital_status  \
0  317991      Lu***    Al*****  29.07.03      AR              1   
1  241892     Ma****  Ro*******  10.11.99      AR              1   
2  303376     Fr****     Ac****  25.07.84      AR              1   
3  761992     Fr****      Go***  20.03.70      AR              1   
4  373318      Jo***      Ro***  05.10.71      AR              1   

   number_of_kids               employment  
0             2.0             Web Designer  
1             0.0                   Banker  
2             0.0            Kitchen staff  
3             0.0       Teaching Assistant  
4             0.0  Administrative employee

#### Column/feature names
print the list of columns in the user records dataset.

In [6]:
#Your code here
list(user_records.columns)

['UserID',
 'first_name',
 'last_name',
 'birthday',
 'country',
 'marital_status',
 'number_of_kids',
 'employment']

#### Duplicates
Check if there are any duplicate entries in the user records dataset.

__Remark__: If there are two rows in the dataset that have the same value for __all of the columns__, we consider this as a duplicate.

In [15]:
#Your code here
user_records.duplicated(keep=False)

0      False
1      False
2      False
3      False
4      False
       ...  
952    False
953    False
954    False
955    False
956    False
Length: 957, dtype: bool

Show all the duplicated rows in the user records dataset.

__Hint:__ use the method `duplicated(keep=False)` to get all occurrences of the duplicated rows.

In [16]:
#Your code here
duplicate=user_records[user_records.duplicated()]
duplicate

UserID first_name last_name  birthday country marital_status  \
19   100684    Ma*****    Me****  17.06.76      AR            NaN   
165  910382      Ni***    El****  21.03.75      AZ            NaN   
290  610518    An*****  An******  15.02.53      DK              1   

     number_of_kids               employment  
19              0.0  Administrative employee  
165             0.0  Administrative employee  
290             NaN                  Manager

**Q2. Are there any duplicate entries in the user records dataset?**

*Note: Duplicate entries refer to two or more rows where all the values across every column are identical.*

Yes, there are 3 duplicate entries.

**Q3. How many duplicate entries are in the user records dataset?**

(Remark: for instance: there are 2 duplicate "entries" in the following list: '1 ,2, 1, 3, 3, 4, 5'. It's 1 and 3).

There are 3. They are the entries with the following User ID: 100684, 910382 and 610518

Drop the duplicate entries in the user records dataset.

In [33]:
#Your code here
duplicate_index=duplicate.index
duplicate_index
user_records.drop(duplicate_index, inplace=True)



### Spies often use false identities with fabricated data.

#### This suggests that if certain data points occur with statistically improbable frequency, they might be fabricated.

In this part we are going to verify if there are birthdates that have been reused an unusual amount of times.

In [65]:
user_records.loc[0, "birthday"]

'29.07.03'

Let's first ensure that the birthday column in the user records dataset has the correct format (datetime):


In [73]:
#Check the data type of each feature
#Your code here
from datetime import datetime


format="%d.%m.%y"
mylist=user_records["birthday"]
for i in range(len(mylist)):
  test_str=mylist[i]
res=True
try:
  res=bool(datetime.strptime(test_str, format))
except ValueError:
    res=False

print(str(res))


KeyError: 19

Convert the column `birthday` to datetime if necessary. Make sure to pass the correct format when parsing dates. In this case, the correct date format is `"%d.%m.%y"`.

In [ ]:
#Your code here


Find the list of birthdays that are duplicated/reused in the dataset

In [ ]:
#Your code here


Let's count how many people have their birthday on the same day among the duplicated birthdays.

In [ ]:
#Your code here


**Q4. What is the most common birthday in the dataset?**

**Q5. How many users have their birthdays on that day (the most common birthday)?**

We can consider that there is no suspicious discrepancy through birthdays.


#### We have reasons to believe that the suspect works very thoroughly, so he would make sure that the fabricated data about his identity is complete in order for his operation to run smoothly.
#### Therefore, we can exclude the people with incomplete information from the list of potential suspects.

Count how many null values are there in each column of the user records dataset. You can call `isnull()` and `sum()` to get a count of how many null values are there in each column.

In [ ]:
#Your code here


If you encounter any null values in the dataset, be sure to exclude those entries from the suspect list.

In [ ]:
#Your code here


### Linking the user records to the flight records

We will now proceed to identify the users who were in the USA from Sept 1st, 2019 to Oct 31st, 2020.

We will merge the flight records with the user records in order to ensure that users who have been unsuspected in the previous step are not considered.

In [ ]:
#check the user records dataset

In [ ]:
#check the flight records dataset


[Merging](https://pandas.pydata.org/docs/user_guide/merging.html) the two datasets:

*Hint: Use the above questions to identify on what feature to merge the datasets*

The result should contain the features flightName, Departure, Arrival,  Date, UserID, first_name, last_name, etc.

In [ ]:
#Your code here


Remember, our suspect operates meticulously, ensuring that there are no null values in any entries.

In [ ]:
#Your code here


Let's now identify the users who has **travelled to** the United States (US) between the 1st of september 2019 and the 31st of october 2020.

*Note that the travel date is not in datetime format. Again, use the date format `"%d.%m.%y"` to do the conversion to datetime.*

In [ ]:
#Your code here


**Q6. How many suspects travelled to the US between the 1st of September 2019 and the 31st of October 2020?**

**Q7. Which users are included in the remaining list of suspects who traveled to the US between the 1st of September 2019 and the 31st of October 2020?** Choose from the answers on Moodle.

## [Data visualisation ](https://pandas.pydata.org/docs/user_guide/visualization.html)

 Suisse Impossible Mission Force is happy with the result and would like to have some visual aid to make a profile of the potential suspects.

Let's create a pie chart to visualize the employment distribution among the suspects who traveled to US between September 1st, 2019, and October 31st, 2020.

In [ ]:
#Your code here


**Q8. Which profession is the least represented among the suspects the suspects who traveled to US between September 1st, 2019, and October 31st, 2020?**
 Choose from the answers on Moodle.


## Journey length

The Suisse Impossible Mission Force suspects that the rogue agent is an anomaly within his declared employment category, particularly in terms of the number of days he spent at destinations compared to his peers.

To identify such deviations, we can employ boxplots for each employment category, allowing us to visually identify outliers in the data.

First, let's construct two tables: one for the dates of arrival and another for the dates of departure. Utilize the existing dataset to extract these dates based on the 'destination' and 'departure' columns. Ensure that both the arrival and departure dates are within the period the suspect is believed to have been in the US, which is from September 2019 to October 2020.

In [ ]:
# For this step we will need the information not only about the fights to the US,
# but also about the flights taken from the US
# Your code here


We can keep on both of these tables the user ID, employment, and departure date / arrival date.

In [ ]:
#Your code here


We can now join the two tables on user ID.

In [ ]:
#Your code here

We can now create a new feature called journey_length representing the difference between departure date and arrival date.

*Hint: Ensure that journey_length is of type numeric as the box plot in the next question requires a numeric type. To convert from datetime to number of days, you can use `dt.days`*

In [ ]:
#Your code here


*Sidenote: if you get negative journey length, what would be your explanation to that?*

We can now create the boxplots where the x-axis represents each profession and the y-axis corresponds to the journey length.

In [ ]:
#Your code here


**Q9. Which profession has the most variability?**

### SIMF is satisfied with your work but believes the current data is inconclusive.

Your investigation will continue when more intelligence comes in...

**Don't forget to complete the moodle quiz ([Assignment Q1](https://moodle.unil.ch/mod/quiz/view.php?id=1744562)) and submit your code ([Code Q1](https://moodle.unil.ch/mod/assign/view.php?id=1744563)) on Moodle before Monday 30.09.2024.**